workflows/cmgg-preprocessing.nf

```diff
@@ -67,56 +66,147 @@ include { MULTIQC                     } from "../modules/nf-core/modules/multiqc
+        .dump(tag: "MAIN: reads inputs"   , {FormattingService.prettyFormat(it)})
+    ch_inputs.other
+        .dump(tag: "MAIN: other inputs"   , {FormattingService.prettyFormat(it)})
+
+    /*
+    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+    // PROCESS FLOWCELL INPUTS
+    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+    */
 
     ch_flowcell = ch_inputs.flowcell.multiMap { meta, samplesheet, flowcell, sample_info ->
         fc   : [meta, samplesheet, flowcell]
         info : sample_info
     }
 
-    // DEMULTIPLEX([meta, samplesheet, flowcell])
-    DEMULTIPLEX(ch_flowcell.fc)
-    DEMULTIPLEX.out.bclconvert_fastq.dump(tag: "bclconvert_fastq",{prettyDump(it)})
-    ch_multiqc_files = ch_multiqc_files.mix(DEMULTIPLEX.out.bclconvert_reports.map { meta, reports -> return reports} )
-    ch_versions      = ch_versions.mix(DEMULTIPLEX.out.versions)
+    // BCL_DEMULTIPLEX([meta, samplesheet, flowcell], demultiplexer)
+    BCL_DEMULTIPLEX(ch_flowcell.fc, "bclconvert")
+    BCL_DEMULTIPLEX.out.fastq.dump(tag: "DEMULTIPLEX: fastq",{FormattingService.prettyFormat(it)})
+    ch_multiqc_files = ch_multiqc_files.mix(
+        BCL_DEMULTIPLEX.out.reports.map { meta, reports -> return reports},
+        BCL_DEMULTIPLEX.out.stats.map   { meta, stats   -> return stats  }
+    )
+    ch_versions = ch_versions.mix(BCL_DEMULTIPLEX.out.versions)
 
     // Add metadata to demultiplexed fastq's
     ch_demultiplexed_fastq = merge_sample_info(
-        DEMULTIPLEX.out.bclconvert_fastq,
+        BCL_DEMULTIPLEX.out.fastq,
         parse_sample_info_csv(ch_flowcell.info)
     )
 
+    /*
+    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+    // PROCESS BAM/CRAM INPUTS
+    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+    */
+
+    // Convert bam/cram inputs to fastq
+    BAM_TO_FASTQ(ch_inputs.reads, ch_fasta_fai)
+    ch_converted_fastq = BAM_TO_FASTQ.out.fastq
+    ch_converted_fastq.dump(tag: "converted_fastq",{FormattingService.prettyFormat(it)})
+
+
+    /*
+    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+    // GATHER PROCESSED INPUTS
+    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+    */
+
     // "Gather" fastq's from demultiplex and fastq inputs
     ch_sample_fastqs = Channel.empty()
-    ch_sample_fastqs = ch_sample_fastqs.mix(ch_inputs.fastq, ch_demultiplexed_fastq)
+    ch_sample_fastqs = ch_sample_fastqs.mix(ch_inputs.fastq, ch_demultiplexed_fastq, ch_converted_fastq)
+
+    /*
+    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+    // FASTQ TRIMMING AND QC
+    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+    */
 
-    //*
-    // STEP: FASTQ TRIMMING AND QC
-    //*
     // MODULE: fastp
```

Comment:
@nvnieuwk, do you see a way here to do some groupKey magic?

Reply:
Starting @ line 214 of 79964b1

Reply:
The main problem here is finding a way to correctly get the group sizes here right?

Reply:
Yes, but we can only get the group size AFTER the groupTuple...

Reply:
I think you'll have to write a little function on the input data that can calculate the suspected sizes of this channel: `ch_sample_fastqs = ch_sample_fastqs.mix(ch_inputs.fastq, ch_demultiplexed_fastq, ch_converted_fastq)`

Reply:
Too much work before the holiday -_- , I'll do it after then

Reply:
Yep it's pretty tricky :D here is an example of how I counted the family sizes, maybe it can help you: https://github.com/CenterForMedicalGeneticsGhent/nf-cmgg-germline/blob/1aa03662ded54be68d4e036a7d061df2793a370c/workflows/cmgg-germline.nf#L263-L290